File: applovin_max/ios/Classes/AppLovinMAXAdViewFactory.m
//
//  AppLovinMAXAdViewFactory.m
//  applovin_max
//
//  Created by Thomas So on 7/17/22.
//

#import "AppLovinMAXAdViewFactory.h"
#import "AppLovinMAXAdView.h"
#import "AppLovinMAX.h"
#import <AppLovinSDK/AppLovinSDK.h>

@interface AppLovinMAXAdViewFactory()
@property (nonatomic, strong) id<FlutterBinaryMessenger> messenger;
@end

@implementation AppLovinMAXAdViewFactory

- (instancetype)initWithMessenger:(id<FlutterBinaryMessenger>)messenger
{
    self = [super init];
    if ( self )
    {
        self.messenger = messenger;
    }
    return self;
}

- (id<FlutterMessageCodec>)createArgsCodec
{
    return [FlutterStandardMessageCodec sharedInstance];
}

- (id<FlutterPlatformView>)createWithFrame:(CGRect)frame viewIdentifier:(int64_t)viewId arguments:(id _Nullable)args
{
    // Ensure plugin has been initialized
    ALSdk *sdk = AppLovinMAX.shared.sdk;
    if ( !sdk )
    {
        [AppLovinMAX log: @"Failed to create MaxAdView widget - please ensure the AppLovin MAX plugin has been initialized by calling 'AppLovinMAX.initialize(...);'!"];
        return nil;
    }
    
    NSString *adUnitId = args[@"ad_unit_id"];
    NSString *adFormatStr = args[@"ad_format"];
    MAAdFormat *adFormat = [adFormatStr isEqualToString: @"mrec"] ? MAAdFormat.mrec : DEVICE_SPECIFIC_ADVIEW_AD_FORMAT;
    
    [AppLovinMAX log: @"Creating MaxAdView widget with Ad Unit ID: %@", adUnitId];
    
    // Optional params
    BOOL isAutoRefreshEnabled = ((NSNumber *) args[@"is_auto_refresh_enabled"]).boolValue; // Defaults to true
    NSString *placement = [args[@"placement"] isKindOfClass: [NSString class]] ? args[@"placement"] : nil; // May be NSNull
    NSString *customData = [args[@"custom_data"] isKindOfClass: [NSString class]] ? args[@"custom_data"] : nil; // May be NSNull
    NSDictionary *extraParameters = [args[@"extra_parameters"] isKindOfClass: [NSDictionary class]] ? args[@"extra_parameters"] : nil;
    NSDictionary *localExtraParameters = [args[@"local_extra_parameters"] isKindOfClass: [NSDictionary class]] ? args[@"local_extra_parameters"] : nil;
    
    return [[AppLovinMAXAdView alloc] initWithFrame: (CGRect) { .size = adFormat.size }
                                             viewId: viewId
                                           adUnitId: adUnitId
                                           adFormat: adFormat
                               isAutoRefreshEnabled: isAutoRefreshEnabled
                                          placement: placement
                                         customData: customData
                                    extraParameters: extraParameters
                               localExtraParameters: localExtraParameters
                                          messenger: self.messenger
                                                sdk: sdk];
}

@end

File: applovin_max/android/src/main/java/com/applovin/applovin_max/AppLovinMAXAdViewFactory.java
package com.applovin.applovin_max;

import android.content.Context;

import com.applovin.mediation.MaxAdFormat;
import com.applovin.sdk.AppLovinSdk;

import java.util.Map;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import io.flutter.plugin.common.BinaryMessenger;
import io.flutter.plugin.common.StandardMessageCodec;
import io.flutter.plugin.platform.PlatformView;
import io.flutter.plugin.platform.PlatformViewFactory;

/**
 * Created by Thomas So on July 17 2022
 */
public class AppLovinMAXAdViewFactory
        extends PlatformViewFactory
{
    private final BinaryMessenger messenger;

    public AppLovinMAXAdViewFactory(final BinaryMessenger messenger)
    {
        super( StandardMessageCodec.INSTANCE );

        this.messenger = messenger;
    }

    @NonNull
    @Override
    public PlatformView create(@Nullable final Context context, final int viewId, final Object args)
    {
        Map<String, Object> params = (Map<String, Object>) args;

        String adUnitId = (String) params.get( "ad_unit_id" );
        String adFormatStr = (String) params.get( "ad_format" );
        MaxAdFormat adFormat = "mrec".equals( adFormatStr ) ? MaxAdFormat.MREC : AppLovinMAX.getDeviceSpecificBannerAdViewAdFormat( context );

        // Optional params
        Integer ad_view_id = params.containsKey( "ad_view_id" ) ? (Integer) params.get( "ad_view_id" ) : null;
        int adViewId = ad_view_id != null ? ad_view_id : 0;
        boolean isAdaptiveBannerEnabled = Boolean.TRUE.equals( params.get( "is_adaptive_banner_enabled" ) ); // Defaults to true
        boolean isAutoRefreshEnabled = Boolean.TRUE.equals( params.get( "is_auto_refresh_enabled" ) ); // Defaults to true
        String placement = params.containsKey( "placement" ) ? (String) params.get( "placement" ) : null;
        String customData = params.containsKey( "custom_data" ) ? (String) params.get( "custom_data" ) : null;
        Map<String, Object> extraParameters = params.containsKey( "extra_parameters" ) ? (Map<String, Object>) params.get( "extra_parameters" ) : null;
        Map<String, Object> localExtraParameters = params.containsKey( "local_extra_parameters" ) ? (Map<String, Object>) params.get( "local_extra_parameters" ) : null;

        return new AppLovinMAXAdView( viewId, adUnitId, adViewId, adFormat, isAdaptiveBannerEnabled, isAutoRefreshEnabled, placement, customData, extraParameters, localExtraParameters, messenger, context );
    }
}
